Count-min Sketch prototype

isamaru opened the pull request:
First iteration of Count-min Sketch prototype

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -1,9 +1,10 @@
 
 def hash_key(i, key):
-    return hash(str(i) + str(key))
+    return xxhash.xxh64(str(key), seed=i).intdigest()
```
Better use one of the standard hash libraries that I linked to in the beginning. I never heard of `xxhash` (doesn't mean it's bad, just maturity/support unknown).

But can you explain why we use custom hashing at all? And why the previous code has the intermediate step of converting to `str`? What's wrong with the built-in `hash`?

Function needs a docstring (what is `i`? not a good variable name)

isamaru replied:
Which links to hash libraries are you referring to?
I was going to use [pyhash](https://github.com/flier/pyfasthash), but I had trouble installing it on both Windows and Unix. 
We can use any good hash function as long as it is deterministic and cheaply seedable.

Python's hash is deliberately initialized with a random seed so that dictionary attacks are impossible. If we used that, we couldn't pickle and reuse stored dictionaries.
Furthermore, we want some fast library implemented in C to skip python altogether.

Will document better.

piskvorky commented on bounded_counter/__init__.py:
```diff
@@ -0,0 +1 @@
+from .count_min_sketch import *
```
Missing newline (as github shows).

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -0,0 +1,99 @@
+algorithm_logcounter = 2
+
+class CountMinSketch(object):
+    def __init__(self, width=16, depth=8, algorithm = None):
```
PEP8: no space around default argument parameters.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -0,0 +1,99 @@
+        array_type = numpy.uint32
+        self.algorithm = algorithm_basic
+        self.inc_method = self.basic_increment
+        if algorithm and str(algorithm).lower() == 'conservative':
```
`algorithm` should have a clear type (such as str, or unicode). No tests, conversions and postprocessing needed or desired. Python is a  duck-typed language.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -0,0 +1,99 @@
+        self.algorithm = algorithm_basic
+        self.inc_method = self.basic_increment
+        if algorithm and str(algorithm).lower() == 'conservative':
+            self.algorithm = algorithm_conservative
```
If this is the case, better use code injection: let the `algorithm` parameter be a function, not a string. More flexible and cleaner.

isamaru replied:
This doesn't support injectable algorithms. The user can merely choose from available implementations. I'm considering switching to several different implementations possibly using inheritance instead of choosing using input parameter.
What do you think?

piskvorky replied:
What's popular in libraries is to allow the user to choose from among a few pre-defined algorithms defined as strings (`algorithm="conservative"` etc), plus allow injecting their object/function in the parameter, for flexibility.

It depends how tightly coupled the algorithms are to `CountMinSketch`. If the interface for `algorithm` is simple (just a few methods, or even a single function?),  if `algorithm` is well encapsulated, then injection is the best choice: `def __init__(..., algorithm=algorithm_conservative)`.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -0,0 +1,99 @@
+        return self.hash_mask & hash_key(i, key)
+
+    def increment(self, key):
+        self.inc_method(key)
```
This method does nothing -- why is it here?

isamaru replied:
inc_method is a pointer to a method chosen upon class initialization. If I rewrite this to distinct implementations (classes), this will go away.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -0,0 +1,99 @@
+        self.inc_method(key)
+
+    def basic_increment(self, key):
+        for i in range(self.depth):
```
`xrange` more efficient (generator vs an explicit list). Here and elsewhere.

isamaru replied:
I'm using python 3 where xrange == range. Should I switch to 2 instead?

piskvorky replied:
Yes, we support both (2.7+ and 3.4+), using the [six](https://pythonhosted.org/six/) library, the differences are minimal.

The exception would be if we had to use some Python3-specific features (`async`?), but I don't think that's needed here.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -0,0 +1,99 @@
+                self.table[i, bucket] += 1
+            else:
+                mask = (1 << (current//self.logbase - 1)) - 1
+                r = int.from_bytes(os.urandom(4), byteorder='big', signed=False)
```
Is this an optimization? At this point, standard `random`/`randint`preferable.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -0,0 +1,99 @@
+                minimum = current
+            else:
+                minimum = min(minimum, current)
+        if self.algorithm == algorithm_logcounter:
```
Better as a part of API: let `decode()` be called always, even if sometimes a no-op.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -0,0 +1,99 @@
+
+    def __getitem__(self, key):
+        minimum = None
+        for i in range(self.depth):
```
This would be easier to read (and faster) as a comprehension in Python:

```python
minimum = min(self.table[depth, self.bucket(depth, key)] for depth in xrange(self.depth))
```

This will raise for empty inputs (self.depth <= 0), but I think that's desirable, that's really an error state.

piskvorky commented on bounded_counter/tests/count_min_sketch_test.py:
```diff
@@ -0,0 +1,25 @@
+import bounded_counter
+from timeit import default_timer as timer
+import cProfile
```
Output of a  line profiler (~kernprof) might be more illuminating for profiling hotspots.

isamaru commented:
I've done some testing using current implementation of BoundedCounter (Count-min sketch) using wikipedia.

I've counted all the bigrams and unigrams exactly (using disk for bigrams, and a python dictionary for unigrams).
Then I've filled a count-min sketch structure using width 67 108 864 (2^26), depth 4 and logcounter1024. The ratio of different bigrams (179 413 989) to width is 2.67:1

Finally, I've used scoring from the phraser algorithm (using min_count=50 and threshold=10.0) and compared the exact score (using exact bigram count and exact unigram count) with the estimated score (using the count-min sketch structure and the exact unigram count). 

This is the summary:

Processed 179413989 bigrams
179403089 ok, 316920 true positives, 179086169 true negatives, 10853 false positives, 47 false negatives
Precision 0.966889, Recall 0.999852, F1: 0.983094

In the attached file you can view the miscategorized bigrams with their scores.

[compare_10.txt](https://github.com/RaRe-Technologies/bounded_counter/files/1261018/compare_10.txt)

piskvorky commented:
That looks great! Awesome results. Too bad for `škoda favorit` and "english literature", but the errors seem understandable and all fairly close.

Can you add some info on:
* the structure size in MB (the main matrix size)
* the same analysis for different sizes (different ratio)

We'll need some hints as to what error levels should the users expect, for a given MB size and ratios (the ratios are always expected rough estimates -- true bigram count is unknown).

Btw, can this structure also report the number of unique elements in it?

isamaru commented:
Good news! I was able to come up with a conservative-update equivalent for the probabilistic counter version. This offers better precision for the counter, and particularly for the low values (where it's most significant)! I'm running tests on Wiki data set now, will post results when I have them.

I also added hyperloglog from [here](https://github.com/ascv/HyperLogLog). That is used to report the number of distinct elements in the structure.

Finally, I've added some additional documentation (readme).

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -2,6 +2,7 @@
 
 import numpy
 import xxhash
+from HLL import HyperLogLog
```
This looks very tricky. Supporting a 3rd party package (especially one in C) is always a maintenance headache. Its support will be on us in reality -- can we vouch for the code / its portability / support?

How difficult is it to implement HyperLogLog "natively"?

isamaru replied:
> Its support will be on us in reality
Yeah, I agree. In this case, it's simple enough, and seems solid.

What do you mean, 'natively'? I deliberately chose the C impl out of multiple options because it's 20x faster than an alternative pure-python [hyperloglog](https://github.com/svpcom/hyperloglog) that I've tried.

Both of them are pretty canonical too: I've checked the source and I can't imagine doing anything differently, maybe apart from choosing a different hash function (and I love this one's use of MMH)

isamaru replied:
We could copy-paste the C code and bundle internally (if that's what you prefer), the license allows that.
In fact, that would directly give us access to the C MMH that I could use in our cythonized Count-min Sketch :)

piskvorky replied:
Yes, that's much better :)

By "natively" I meant within our control, without relying on 3rd party code and release cycles.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -136,10 +142,10 @@ def log_encode(self, input):
+        return value
+
+    def __len__(self):
+        return int(self.distinct_counter.cardinality())
```
Is this only approximative, right?

If so, there has to be a big fat warning (or perhaps even a differently named method, Python has a clear contract for `__len__` and things may break in unexpected places).

isamaru replied:
Sure, let's do just "cardinality"!

piskvorky commented on README.md:
```diff
@@ -1 +1,50 @@
+    For significantly smaller widths, deterioration will occur.
+    For instance, to store all bigrams in English Wikipedia (1 857 420 106 bigrams, 179 413 989 unique),
+    good results can be achieved with a width of 67 108 864 (2^26) (37%).
+-   **Depth**: Number of rows, significant for the reliability of the result. Linearly affects speed of the
```
Way too complicated. The API should accept a memory limit, in (mega/giga)bytes.

We can keep these width/depth params exposed, for power users, but not by default.

menshikh-iv commented on bounded_counter/count_min_sketch.py:
```diff
@@ -0,0 +1,152 @@
+    """
+    Data structure used to estimate frequencies of elements in massive data sets with fixed memory footprint.
+    To calculate memory footprint:
+        ( width * depth * cell_size ) + HLL size
```
Need static method for size calculation.

menshikh-iv commented on README.md:
```diff
@@ -1 +1,50 @@
+   - 2B for logcounter1024 / logcons1024
+HLL size is 32 KB
+
+Example:
```
Need code examples (how to use it) and benchmark results.

isamaru commented on bounded_counter/count_min_sketch.py:
```diff
@@ -0,0 +1,152 @@
+        self.distinct_counter = HyperLogLog(15)
+
+    def bucket(self, i, key):
+        return self.hash_mask & hash_key(i, key)
```
Yeah, no problem. I don't expect mask to be larger than 32 bytes but it's nice to have that luxury. I consider the 64-bit hash to be a little better (allows better perturbation). I had already tried performance of both 64 and 32 bit versions and the difference is negligible.

piskvorky replied:
Do you mean 32 bits?

If we have assumptions or known invariants, we should reflect them in an explicit `assert` (or at least a clear comment).

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -17,71 +17,148 @@ def hash_key(row, key):
+            print(cms.depth)   # 8
+            print(cms.size)  # 536 870 912 (512 MB in bytes)
+
+            cms.increment("foo")
```
We agreed on Counter interface + raising `NotImplementedError` for methods that cannot be done using a given structure (key iteration etc).

Or is the left for the final "wrapper", which will use this `CountMinSketch` class internally? (=this class not directly meant for end users?)

isamaru replied:
For CountMin-Sketch, trying to mimic the Counter interface doesn't really make sense. Even supporting the most traditional `cms['foo']+=1` is not possible(1) due to how python interprets such call (always as a successive __getitem__ -> __setitem__).

(1)You *could* work around that by writing a wrapper which returns custom objects instead of integers that delegate their __iadd__ to the structure but that sounds like a major overkill.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -17,71 +17,148 @@ def hash_key(row, key):
         """
         Initializes the Count-Min Sketch structure with the given parameters
-        :param width (int) controls the number of hash buckets in one row of the table, should always be power of 2.
+        :param size_mb (int) controls the maximum size of the Count-min Sketch table.
```
We use Google-style docstrings (I sent a link earlier).

menshikh-iv replied:
But google-style only for gensim, re-structured text is nice I think.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -17,71 +17,148 @@ def hash_key(row, key):
+
+Example:
+        {
+            cms = CountMinSketch(512) # Use 512 MB
```
Named argument better: `size_mb=512` (or `max_size_mb=512`, to be clearer).

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -17,71 +17,148 @@ def hash_key(row, key):
     """
     Data structure used to estimate frequencies of elements in massive data sets with fixed memory footprint.
+
+Example:
```
Bad indent.

piskvorky commented on bounded_counter/count_min_sketch.py:
```diff
@@ -17,71 +17,148 @@ def hash_key(row, key):
     Data structure used to estimate frequencies of elements in massive data sets with fixed memory footprint.
+
+Example:
+        {
```
What is this `{`?

Please use restructured text (.rst) for literal (code) formatting in docstrings.

In markdown files (.md), use Markdown (```python ...).
